### PR TITLE
Added JSON Customization Support For Angle Fields

### PR DESCRIPTION
**Pull request by BeksOmega**
## The basics

- [X] I branched from develop
- [X] My pull request is against develop
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

#2315 plus some

### Proposed Changes

* Adds modes: Compass and Protractor to the angle field JSON definition.
* Also adds properties: clockwise, offset, wrap, and round to the angle field JSON definition.
* Adds functions for setting the individual properties programmatically.
* Adds functions for getting the value of the property (with respect for globals).

### Reason for Changes

* Requested.
* More info below.
* Good practice.
* Good practice.

#### Why did I add individual properties?

I absolutely love love love angle fields. I think they're probably the coolest field in Blockly and I think they're super under utilized. I would hate to see them hamstrung by the JSON.

For example, if there were only the two modes available you couldn't make super cool blocks like this:
AngelJSON

which have built-in affordances for directionally-challenged users.

Providing the modes & the properties (with the properties given override power since they're more specific) is the best of both ease of use, and power.

### Test Coverage

Added test blocks for all of the modes and all of the properties.

Tested all of the blocks to make sure they were functioning properly.

Tested on:
* Desktop Chrome

### Additional Information

- [ ] Doc ticket should be created once this is merged.

Setting round to 0 looks crazy lol:
AngelJSON_Round0

But devs can fix it by setting round to 1 instead, so it doesn't need a fix.

### CODE REVIEW

File: core/field_angle.js [hunk 1]
```diff
@@ -72,51 +112,83 @@ Blockly.FieldAngle.fromJson = function(options) {
...
- *   Blockly.FieldAngle.OFFSET = 90;
+/**
+ * Radius of protractor circle.  Slightly smaller than protractor size since
+ * otherwise SVG crops off half the border at the edges.
+ * @type {number}
+ * @const
  */
+Blockly.FieldAngle.RADIUS = Blockly.FieldAngle.HALF - 1;
+
+/* See <DOCS LINK HERE> for more information about these properties. */
```
Comment (BeksOmega) [inline core/field_angle.js]: Needs a link.

**Comment by BeksOmega** [comment]
Do you want any block factory support for this? Maybe a mutator with the extra options?

File: core/field_angle.js [hunk 2]
```diff
@@ -347,10 +423,104 @@ Blockly.FieldAngle.prototype.doClassValidation_ = function(newValue) {
...
 };
 
+/**
+ * Set which direction should make the graphical angle editor increase.
+ * @param {?boolean} clockwise Whether the graphical angle editor should
+ *    increase as it is moved clockwise (true) or counterclockwise (false)
+ *    or the global setting (null).
+ * @return {!Blockly.FieldAngle} Returns itself (for method chaining).
+ */
+Blockly.FieldAngle.prototype.setClockwise = function(clockwise) {
```
Comment (rachel-fenichel) [inline core/field_angle.js]: @RoboErikG thoughts on returning itself for method chaining?  It's a reasonable pattern, but 1) we don't use it anywhere else and 2) we probably want to be consistent.

Comment (RoboErikG) [reply core/field_angle.js]: It's nice for convenience, but it's pretty non-standard for instances. I think I'd rather start using an opt_config parameter with a map of the configuration options. The keys should be the same as the JSON config which will also let you use the same code for both.

Comment (BeksOmega) [reply core/field_angle.js]: Do you want me to go ahead and update it? or are you waiting for feedback from Rachel?

Comment (rachel-fenichel) [reply core/field_angle.js]: +1 to Erik.  Go ahead and update it.

Comment (BeksOmega) [reply core/field_angle.js]: I went to go update this, and then I realized I had more questions and wasn't sure how far reaching the change should be. I [created an issue](https://github.com/google/blockly/issues/2722) to make discussion easier.

Comment (rachel-fenichel) [reply core/field_angle.js]: If you're happy with the clarification on #2722, then just remove the method chaining and this will be good to go.

Comment (BeksOmega) [reply core/field_angle.js]: I was planning on getting the abstract field changes in first (just to make sure how I'm thinking about it in my head is actually going to work). But don't worry I haven't forgotten about this little guy!

Also atm #2722 has this guy down as no hooks. This is b/c, for example,  changing the WRAP value at runtime means the field would have to do some extra work recalculating it's value. I think I could get it to work though.

Did you want hooks for the angle field or no?

**Comment by BeksOmega** [comment]
Closing after discussion with Sam. Will revive after configuration is happy.

**Pull request closed without merging**